connectors/sources/s3.py

```diff
@@ -187,10 +211,21 @@ async def get_docs(self, filtering=None):
         Yields:
             dictionary: Document from Amazon S3.
         """
+        if self.bucket_list == []:
+            await self.client()
+            s3 = self.s3_region_client["default"]
+            self.bucket_list = await s3.list_buckets()
         bucket_list = self.buckets if self.buckets != ["*"] else self.get_bucket_list()
         page_size = self.configuration["page_size"]
         for bucket in bucket_list:
             region_name = await self.get_bucket_region(bucket)
+            if region_name is not None:
+                await self.client(region=region_name)
+                s3_client = self.s3_region_client[region_name]
```

Comment:
This is a weird pattern. It's having `self.client()` side effect the presence of `self.s3_region_client[region_name]`. 

Instead, we should just have a `self.get_client()` that returns the client, so that you don't have to do this in two lines all over the place.